overlord/snapstate/handlers.go

```diff
@@ -892,113 +886,195 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 			return err
 		}
 
-		return asyncRefreshOnSnapClose(m.state, refreshInfo)
+		return asyncRefreshOnSnapClose(m.state, snapsup, refreshInfo)
 	}
 
-	continueInhibitedAutoRefresh(st)
+	continueInhibitedAutoRefresh(st, snapsup.InstanceName())
 	return nil
 }
 
 // asyncRefreshOnSnapClose asynchronously waits for the snap the close, notifies
 // the user and then triggers an auto-refresh.
-func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSnapRefreshInfo) error {
-	monitoredSnaps := snapMonitoring(st)
-	if monitoredSnaps == nil {
-		monitoredSnaps = make(map[string]chan<- bool)
-	}
-
+func asyncRefreshOnSnapClose(st *state.State, snapsup *SnapSetup, refreshInfo *userclient.PendingSnapRefreshInfo) error {
+	snapName := snapsup.InstanceName()
 	// there's already a goroutine waiting for this snap to close so just notify
-	if monitoredSnaps[refreshInfo.InstanceName] != nil {
+	if abortChan := getAbortMonitoringChan(st, snapName); abortChan != nil {
 		asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
 		return nil
 	}
 
 	// monitor the snap until it closes. Use buffered channel to prevent the sender
 	// from blocking if the receiver stops before reading from it
 	done := make(chan string, 1)
-	if err := cgroupMonitorSnapEnded(refreshInfo.InstanceName, done); err != nil {
+	if err := cgroupMonitorSnapEnded(snapName, done); err != nil {
 		return fmt.Errorf("cannot monitor for snap closure: %w", err)
 	}
 
 	// notify the user about the blocked refresh
 	asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
 
+	// TODO: since we're assuming that the gated auto-refresh is on by default, the
+	// hints should be set. Use the task's snapSetup as fallback anyway?
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		return fmt.Errorf("cannot get refresh-candidates: %v", err)
+	}
+
 	abort := make(chan bool, 1)
-	monitoredSnaps[refreshInfo.InstanceName] = abort
-	updateMonitoringState(st, monitoredSnaps)
+	if err := addMonitoring(st, snapName, abort); err != nil {
+		return fmt.Errorf("cannot save monitoring state: %v", err)
+	}
+
+	go continueRefreshOnSnapClose(st, snapsup, done, abort)
+	return nil
+}
+
+// addMonitoring adds monitoring info to the persisted and in-memory states.
+func addMonitoring(st *state.State, snapName string, abort chan<- bool) error {
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		return fmt.Errorf("cannot get refresh-candidates: %v", err)
+	} else if _, ok := refreshHints[snapName]; !ok {
```

Comment:
I'm missing were we are adding always setting refresh-candidates in the normal auto-refresh path? refreshhints does that but is done on a separate schedule than the actual auto-refresh. Maybe it's done somewhere and I'm not seeing it? otherwise maybe it should be done as its own preparatory PR

Reply:
It's set in the first phase of the gated auto-refresh in `autoRefreshPhase1`. Since this runs after an inhibited auto-refresh, the refresh-candidates were updated recently. Or do you mean in the "old" path without gated auto-refreshes? I thought we were going to enable it by default as part of this fix (since we had to do it anyway) but maybe I got the wrong impression

Reply:
no, we can't just enable it by default for now because it doesn't have the right logic for RAA in general and the interaction between RAA and the hook is sets up, that all needs more work. We need to copy some bits of it into the normal path for now. Sorry I probably wasn't clear as much I thought about this point.

Reply:
Ah I see, then like you mentioned I'll open a PR for that before all the rest